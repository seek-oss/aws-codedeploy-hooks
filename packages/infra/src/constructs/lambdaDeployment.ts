import { Tags, aws_codedeploy, aws_lambda } from 'aws-cdk-lib';
import { Construct } from 'constructs';

import { commit, version } from '../version';

const tagValue = `${version}-${commit}`;

export type LambdaDeploymentProps = {
  lambdaFunction: aws_lambda.Function;
};

export class LambdaDeployment extends Construct {
  alias: Readonly<aws_lambda.Alias>;

  constructor(
    scope: Construct,
    id: string | null,
    { lambdaFunction }: LambdaDeploymentProps,
  ) {
    super(scope, id ?? 'LambdaDeployment');

    Tags.of(lambdaFunction).add('aws-codedeploy-hooks', tagValue);

    const alias = lambdaFunction.addAlias('Live', {
      description: 'The Lambda version currently receiving traffic',
    });

    this.alias = alias;

    const application = new aws_codedeploy.LambdaApplication(
      this,
      'CodeDeployLambdaApplication',
    );

    Tags.of(application).add('aws-codedeploy-hooks', tagValue);

    const deploymentGroup = new aws_codedeploy.LambdaDeploymentGroup(
      this,
      'CodeDeployLambdaDeploymentGroup',
      {
        application,
        alias,
        deploymentConfig: aws_codedeploy.LambdaDeploymentConfig.ALL_AT_ONCE,
      },
    );

    Tags.of(deploymentGroup).add('aws-codedeploy-hooks', tagValue);

    deploymentGroup.addPreHook(
      aws_lambda.Function.fromFunctionName(
        this,
        'LambdaFunctionPreHook',
        'aws-codedeploy-hook-BeforeAllowTraffic',
      ),
    );

    deploymentGroup.addPostHook(
      aws_lambda.Function.fromFunctionName(
        this,
        'LambdaFunctionPostHook',
        'aws-codedeploy-hook-AfterAllowTraffic',
      ),
    );
  }
}
