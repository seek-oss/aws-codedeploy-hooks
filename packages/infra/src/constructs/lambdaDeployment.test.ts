import {
  App,
  Stack,
  assertions,
  aws_lambda,
  aws_lambda_event_sources,
  aws_sqs,
} from 'aws-cdk-lib';

import { LambdaDeployment } from './lambdaDeployment';

jest.mock('../version', () => ({
  commit: 'abcdefg',
  version: '0.0.0',
}));

it('returns expected CloudFormation stack', () => {
  const app = new App();

  const stack = new Stack(app);

  const lambdaFunction = new aws_lambda.Function(stack, 'LambdaFunction', {
    code: aws_lambda.Code.fromInline(
      'export const handler = async () => undefined;',
    ),
    handler: 'index.handler',
    runtime: aws_lambda.Runtime.NODEJS_20_X,
  });

  const deployment = new LambdaDeployment(stack, null, { lambdaFunction });

  const queue = aws_sqs.Queue.fromQueueArn(
    stack,
    'Queue',
    'arn:aws:sqs:us-east-2:123456789012:queue',
  );

  const eventSource = new aws_lambda_event_sources.SqsEventSource(queue);

  deployment.alias.addEventSource(eventSource);

  const template = assertions.Template.fromStack(stack);

  template.resourceCountIs('AWS::Lambda::Alias', 1);
  template.hasResourceProperties('AWS::Lambda::Alias', {
    Name: 'Live',
  });

  template.resourceCountIs('AWS::Lambda::Function', 1);
  template.hasResourceProperties('AWS::Lambda::Function', {
    Tags: [{ Key: 'aws-codedeploy-hooks' }],
  });

  template.resourceCountIs('AWS::CodeDeploy::Application', 1);
  template.hasResourceProperties('AWS::CodeDeploy::Application', {
    ComputePlatform: 'Lambda',
    Tags: [{ Key: 'aws-codedeploy-hooks' }],
  });

  template.resourceCountIs('AWS::CodeDeploy::DeploymentGroup', 1);
  template.hasResourceProperties('AWS::CodeDeploy::DeploymentGroup', {
    AutoRollbackConfiguration: {
      Enabled: true,
      Events: ['DEPLOYMENT_FAILURE'],
    },
    DeploymentConfigName: 'CodeDeployDefault.LambdaAllAtOnce',
    DeploymentStyle: {
      DeploymentOption: 'WITH_TRAFFIC_CONTROL',
      DeploymentType: 'BLUE_GREEN',
    },
    Tags: [{ Key: 'aws-codedeploy-hooks' }],
  });

  const json = JSON.stringify(template.toJSON());

  expect(JSON.parse(json)).toMatchSnapshot();
});
