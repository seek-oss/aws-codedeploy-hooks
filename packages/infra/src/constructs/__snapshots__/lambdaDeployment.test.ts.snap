// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns expected CloudFormation stack 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "LambdaDeploymentCodeDeployLambdaApplication14FFF847": {
      "Properties": {
        "ComputePlatform": "Lambda",
        "Tags": [
          {
            "Key": "aws-codedeploy-hooks",
            "Value": "0.0.0-abcdefg",
          },
        ],
      },
      "Type": "AWS::CodeDeploy::Application",
    },
    "LambdaDeploymentCodeDeployLambdaDeploymentGroupC2B0BCFA": {
      "Properties": {
        "ApplicationName": {
          "Ref": "LambdaDeploymentCodeDeployLambdaApplication14FFF847",
        },
        "AutoRollbackConfiguration": {
          "Enabled": true,
          "Events": [
            "DEPLOYMENT_FAILURE",
          ],
        },
        "DeploymentConfigName": "CodeDeployDefault.LambdaAllAtOnce",
        "DeploymentStyle": {
          "DeploymentOption": "WITH_TRAFFIC_CONTROL",
          "DeploymentType": "BLUE_GREEN",
        },
        "ServiceRoleArn": {
          "Fn::GetAtt": [
            "LambdaDeploymentCodeDeployLambdaDeploymentGroupServiceRole2775DE50",
            "Arn",
          ],
        },
        "Tags": [
          {
            "Key": "aws-codedeploy-hooks",
            "Value": "0.0.0-abcdefg",
          },
        ],
      },
      "Type": "AWS::CodeDeploy::DeploymentGroup",
    },
    "LambdaDeploymentCodeDeployLambdaDeploymentGroupServiceRole2775DE50": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codedeploy.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSCodeDeployRoleForLambdaLimited",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "aws-codedeploy-hooks",
            "Value": "0.0.0-abcdefg",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaDeploymentCodeDeployLambdaDeploymentGroupServiceRoleDefaultPolicyBB5F5248": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":lambda:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":function:aws-codedeploy-hook-BeforeAllowTraffic",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":lambda:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":function:aws-codedeploy-hook-BeforeAllowTraffic:*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":lambda:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":function:aws-codedeploy-hook-AfterAllowTraffic",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":lambda:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":function:aws-codedeploy-hook-AfterAllowTraffic:*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LambdaDeploymentCodeDeployLambdaDeploymentGroupServiceRoleDefaultPolicyBB5F5248",
        "Roles": [
          {
            "Ref": "LambdaDeploymentCodeDeployLambdaDeploymentGroupServiceRole2775DE50",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LambdaFunctionAliasLive400E15E9": {
      "Properties": {
        "Description": "The Lambda version currently receiving traffic",
        "FunctionName": {
          "Ref": "LambdaFunctionBF21E41F",
        },
        "FunctionVersion": {
          "Fn::GetAtt": [
            "LambdaFunctionCurrentVersion7D5C86E3e2b46016d2d80a96993d54a126fb6e88",
            "Version",
          ],
        },
        "Name": "Live",
      },
      "Type": "AWS::Lambda::Alias",
      "UpdatePolicy": {
        "CodeDeployLambdaAliasUpdate": {
          "AfterAllowTrafficHook": {
            "Fn::Select": [
              6,
              {
                "Fn::Split": [
                  ":",
                  {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":lambda:",
                        {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":function:aws-codedeploy-hook-AfterAllowTraffic",
                      ],
                    ],
                  },
                ],
              },
            ],
          },
          "ApplicationName": {
            "Ref": "LambdaDeploymentCodeDeployLambdaApplication14FFF847",
          },
          "BeforeAllowTrafficHook": {
            "Fn::Select": [
              6,
              {
                "Fn::Split": [
                  ":",
                  {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":lambda:",
                        {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":function:aws-codedeploy-hook-BeforeAllowTraffic",
                      ],
                    ],
                  },
                ],
              },
            ],
          },
          "DeploymentGroupName": {
            "Ref": "LambdaDeploymentCodeDeployLambdaDeploymentGroupC2B0BCFA",
          },
        },
      },
    },
    "LambdaFunctionAliasLiveSqsEventSourceQueue453DF2EA": {
      "Properties": {
        "EventSourceArn": "arn:aws:sqs:us-east-2:123456789012:queue",
        "FunctionName": {
          "Fn::Join": [
            "",
            [
              {
                "Fn::Select": [
                  6,
                  {
                    "Fn::Split": [
                      ":",
                      {
                        "Ref": "LambdaFunctionAliasLive400E15E9",
                      },
                    ],
                  },
                ],
              },
              ":Live",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "LambdaFunctionBF21E41F": {
      "DependsOn": [
        "LambdaFunctionServiceRoleDefaultPolicy32EEEE35",
        "LambdaFunctionServiceRoleC555A460",
      ],
      "Properties": {
        "Code": {
          "ZipFile": "export const handler = async () => undefined;",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LambdaFunctionServiceRoleC555A460",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "aws-codedeploy-hooks",
            "Value": "0.0.0-abcdefg",
          },
        ],
      },
      "Type": "AWS::Lambda::Function",
    },
    "LambdaFunctionCurrentVersion7D5C86E3e2b46016d2d80a96993d54a126fb6e88": {
      "Properties": {
        "FunctionName": {
          "Ref": "LambdaFunctionBF21E41F",
        },
      },
      "Type": "AWS::Lambda::Version",
    },
    "LambdaFunctionServiceRoleC555A460": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "aws-codedeploy-hooks",
            "Value": "0.0.0-abcdefg",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaFunctionServiceRoleDefaultPolicy32EEEE35": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:sqs:us-east-2:123456789012:queue",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LambdaFunctionServiceRoleDefaultPolicy32EEEE35",
        "Roles": [
          {
            "Ref": "LambdaFunctionServiceRoleC555A460",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
